src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -128,25 +125,20 @@ extends TyperState(r) {
    */
   override def commit()(implicit ctx: Context) = {
     val targetState = ctx.typerState
-    assert(targetState eq previous)
     assert(isCommittable)
-    targetState.constraint = constraint
+    targetState.constraint =
+      if (targetState.constraint eq previousConstraint) constraint
+      else targetState.constraint & constraint
```

Comment:
The commit message of https://github.com/lampepfl/dotty/pull/1482/commits/36648547105bc8ac53dd63a03821a27243af13cd explains why this is needed but this is tricky enough that it should also be explained in a comment here.